core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -824,8 +824,7 @@ public static void validateStateIsFromCurrentMaster(ESLogger logger, DiscoveryNo
         }
     }
 
-    void handleJoinRequest(final DiscoveryNode node, final MembershipAction.JoinCallback callback) {
-
+    void handleJoinRequest(final DiscoveryNode node, final ClusterState state, final MembershipAction.JoinCallback callback) {
```

Comment:
We can but passing the cluster state as param makes this easily testable. Initially, I tried writing a test with two nodes, one having a custom plugin and another none, but the testing framework makes it tricky. any ideas on how to test this functionality, if we just sample the state instead of passing it as a param?